hydra-node/src/Hydra/Chain/Blockfrost/Client.hs

```diff
@@ -0,0 +1,378 @@
+import Control.Lens ((.~), (^.))
+import Data.Default (def)
+import Data.SOP.NonEmpty (NonEmpty (..))
+import Data.Set qualified as Set
+import Hydra.Cardano.Api.Prelude (StakePoolKey, fromNetworkMagic)
+import Hydra.Chain.CardanoClient (buildTransactionWithPParams')
+import Hydra.Contract.Head qualified as Head
+import Hydra.Ledger.Cardano (adjustUTxO)
+import Hydra.Plutus (commitValidatorScript, initialValidatorScript)
+import Money qualified
+import Ouroboros.Consensus.Block (GenesisWindow (..))
+import Ouroboros.Consensus.Cardano.Block (CardanoEras)
+import Ouroboros.Consensus.HardFork.History (EraEnd (..), EraParams (..), EraSummary (..), SafeZone (..), Summary (..), initBound, mkInterpreter)
+
+data APIBlockfrostError
+  = BlockfrostError Text
+  | DecodeError Text
+  deriving (Show, Exception)
+
+runBlockfrostM ::
+  (MonadIO m, MonadThrow m) =>
+  Blockfrost.Project ->
+  BlockfrostClientT IO a ->
+  m a
+runBlockfrostM prj action = do
+  result <- liftIO $ runBlockfrost prj action
+  case result of
+    Left err -> throwIO (BlockfrostError $ show err)
+    Right val -> pure val
+
+publishHydraScripts ::
+  -- | The path where the Blockfrost project token hash is stored.
+  FilePath ->
+  -- | Keys assumed to hold funds to pay for the publishing transaction.
+  SigningKey PaymentKey ->
+  IO [TxId]
+publishHydraScripts projectPath sk = do
+  prj <- Blockfrost.projectFromFile projectPath
+  runBlockfrostM prj $ do
+    genesis@Blockfrost.Genesis
+      { _genesisNetworkMagic = networkMagic
+      , _genesisSystemStart = systemStart'
+      } <-
+      Blockfrost.getLedgerGenesis
+    pparams <- toCardanoPParams
+    let address = Blockfrost.Address (vkAddress networkMagic)
+    let networkId = toCardanoNetworkMagic networkMagic
+    let changeAddress = mkVkAddress networkId vk
+    stakePools' <- Blockfrost.listPools
+    let stakePools = Set.fromList (toCardanoPoolId <$> stakePools')
+    let systemStart = SystemStart $ posixSecondsToUTCTime systemStart'
+    let eraHistory = mkEraHistory genesis
+    utxo <- Blockfrost.getAddressUtxos address
+    let cardanoUTxO = toCardanoUTxO utxo changeAddress
+    flip evalStateT cardanoUTxO $
+      forM scripts $ \script -> do
+        nextUTxO <- get
+        let output = mkScriptTxOut pparams networkId <$> [mkScriptRef script]
+            totalDeposit = sum (selectLovelace . txOutValue <$> output)
+            someUTxO =
+              maybe mempty UTxO.singleton $
+                UTxO.find (\o -> selectLovelace (txOutValue o) > totalDeposit) nextUTxO
+        (tx, body) <-
+          liftIO $
+            buildTransactionWithPParams' pparams systemStart eraHistory stakePools changeAddress someUTxO [] output
+              >>= \case
+                Left err -> throwErrorAsException err
+                Right rawTx -> do
+                  let body = getTxBody rawTx
+                  pure (makeSignedTransaction [makeShelleyKeyWitness body (WitnessPaymentKey sk)] body, body)
```

Comment:
Should: not duplicate the logic for creating the script transactions

We are using transaction chaining here, good. We should be using the same technique for the direct cardano node connection. The only difference then would be how we query things like system start, pparams etc.